inductor cpp wrapper: add GIL release and acquire

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111888

Support multiple instances inference (in different threads of the same process) as in https://github.com/pytorch/pytorch/issues/93524#issuecomment-1421816158.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

chunyuan-w committed:
inductor cpp wrapper: add GIL release and acquire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release and acquire"


Support multiple instances inference (in different threads of the same process) as in https://github.com/pytorch/pytorch/issues/93524.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1427,6 +1433,7 @@ def generate_return(self, output_refs):
                         + f"new at::Tensor({output}));"
                     )
         else:
+            self.wrapper_call.writeline("py::gil_scoped_acquire acquire;")
```
with scoped release at the beginning of the function, you don't need to explicitly acquire the gil here, right?

chunyuan-w replied:
Removed the explicit acquire here.

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release and acquire"


Support multiple instances inference (in different threads of the same process) as in https://github.com/pytorch/pytorch/issues/93524#issuecomment-1421816158.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire reviewed (commented):
Test case?

jansel reviewed (commented):
Is it currently possible for cpp wrapper to call into python code?

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release and acquire"


Support multiple instances inference (in different threads of the same process) as in https://github.com/pytorch/pytorch/issues/93524#issuecomment-1421816158.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 commented:
> Is it currently possible for cpp wrapper to call into python code?

I'm not aware of a path in cpp wrapper to call into python code? Any concern about it?

chunyuan-w commented:
> Test case?

I added a test case to cover the functionality of a multi-threading scenario.

chunyuan-w commented:
@pytorchbot rebase

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release and acquire"


Support multiple instances inference (in different threads of the same process) as in https://github.com/pytorch/pytorch/issues/93524#issuecomment-1421816158.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge

chunyuan-w commented:
@pytorchbot merge
